File: spdeBarrier/spdeBarrier.R
#Currently needs a version of TMB which includes the barrier procedure
#devtools::install_github("OlavNikolaiBreivik/adcomp/TMB", ref = "barrierAndAD")

library(INLA)
library(fields)
library(TMB)

#Compile and load c-code
compile("spdeBarrier.cpp")
dyn.load("spdeBarrier")

#Download and read data
download.file(url = "https://haakonbakka.bitbucket.io/data/WebSiteData-Archipelago.RData", destfile = "WebSiteData-Archipelago.RData")
load(file = "WebSiteData-Archipelago.RData")

#Construtc mesh
mesh = inla.mesh.2d(boundary = poly.water,
                    loc=cbind(df$locx, df$locy),
                    max.edge = c(1,5)*0.6,
                    cutoff = 0.06,
                    offset = c(0.6, 4.6))
#Plot mesh
png(file = "mapAndMesh.png",height = 900,width = 900)
plot(mesh) 
points(df$locx, df$locy, col="red",cex = sqrt(df$y.smelt)+1)
dev.off()

#Construct spatial interpolation matrix
A = inla.spde.make.A(mesh, loc = cbind(df$locx, df$locy))


#Construct barrier.triangles
tl = length(mesh$graph$tv[,1])
posTri = matrix(0, tl, 2)
for(t in 1:tl){
  temp = mesh$loc[mesh$graph$tv[t, ], ]
  posTri[t,] = colMeans(temp)[c(1,2)]
}
posTri = SpatialPoints(posTri)
normal = unlist(over(poly.water, posTri, returnList=T))
barrier.triangles = setdiff(1:tl, normal)


#Construct spde object and structure needed in the spde-procedure
spde = inla.spde2.matern(mesh, alpha=2)
spdeMatrices = spde$param.inla[c("M0","M1","M2")] #Matrices needed in Standard spde-procedure
fem = INLA:::inla.barrier.fem(mesh = mesh, barrier.triangles = barrier.triangles)
spdeMatricesBarrier = list(C0 = fem$C[[1]],C1 =fem$C[[2]] ,D0 = fem$D[[1]],D1 = fem$D[[2]],I = fem$I )



#Construct design matrix for fixed effects
#X <- model.matrix( ~ 1 + dptLUKE + dptavg15km + dist30m + joetdsumsq + lined15km + swmlog10 + temjul15,
#                   data = df) #Use covariates
X <- model.matrix( ~ 1,
                   data = df) #No covariates



#Define data and parameters-------
data = list(y = df$y.smelt,
            A = A,
            spdeMatrices = spdeMatrices,
            spdeMatricesBarrier = spdeMatricesBarrier,
            barrier = 1,
            c = c(1,0.2),
            X = as.matrix(X)
)

par = list(beta = rep(0,dim(X)[2]),
           log_tau =1,
           log_kappa = -3,
           log_sigmaIID = -3,
           x = rep(0,mesh$n),
           xiid = rep(0,length(df$y.smelt)))
#--------------------------------

#Estimating the model and extract results-------------
startTime <- Sys.time()
obj <- MakeADFun(data, par, random=c("x", "xiid"), DLL="spdeBarrier")
opt <- nlminb(obj$par, obj$fn, obj$gr)
rep = sdreport(obj)
endTime = Sys.time()
timeUsed = endTime - startTime
print(timeUsed)
#------------------------

#Extract range
rangeIndex = which(row.names(summary(rep,"report"))=="range")
range = summary(rep,"report")[rangeIndex,]

#Plot spatial effect
png("spatialBarrierTMB.png")
proj = inla.mesh.projector(mesh)
latentFieldMAP = rep$par.random[names(rep$par.random)=="x"]/exp(rep$par.fixed[which(names(rep$par.fixed)=="log_tau")])
image.plot(proj$x,proj$y, inla.mesh.project(proj, latentFieldMAP),col =  colorRampPalette(c("white","yellow", "red"))(12),
           xlab = 'Easting', ylab = 'Northing',zlim = c(-4,4),
           main = "MAP estimate of spatial latent field",
           cex.lab = 1.1,cex.axis = 1.1, cex.main=1, cex.sub= 1.1)
plot(inla.barrier.polygon(mesh, barrier.triangles), add=T)
contour(proj$x, proj$y,inla.mesh.project(proj, latentFieldMAP) ,add = T,labcex  = 1,cex = 1)
dev.off()

File: spdeBarrier/spdeBarrier.cpp
#include <TMB.hpp>
template<class Type>
  Type objective_function<Type>::operator() ()
{
  using namespace R_inla; //includes SPDE-spesific functions, e.g. Q_spde()
  using namespace density;
  using namespace Eigen; //Needed for utilisation of sparse structures

  //Load data and parameters----------------
  DATA_VECTOR(y);      //The response
  DATA_STRUCT(spdeMatrices,spde_t); //Three matrices needed for representing the GMRF, see p. 8 in Lindgren et al. (2011)
  DATA_STRUCT(spdeMatricesBarrier,spde_b); //Structure needed for the barrier procedure
  DATA_SPARSE_MATRIX(A);  //Matrix for interpolating points witin triangles
  DATA_INTEGER(barrier); // if barrier==1, barrier procedure is used
  DATA_VECTOR(c);      // Scaling of range
  DATA_MATRIX(X);         //Design matrix for fixed effects
  
  

  PARAMETER_VECTOR(beta); //Regresion coefficients for intercept and covariates
  PARAMETER(log_tau); //log precision for spatial effect
  PARAMETER(log_kappa); //log spatial scale parameter in spatial effect
  PARAMETER(log_sigmaIID); //log standard deviation in iid effect
  PARAMETER_VECTOR(x); //Spatial effect
  PARAMETER_VECTOR(xiid);  //Iid effect
  
  //---------------------------------------

  //Transform parameters-------------------
  Type tau = exp(log_tau);
  Type kappa = exp(log_kappa);
  Type sigmaIID = exp(log_sigmaIID);
  //------------------------------------------

  // Spatial interpolation
  vector<Type> delta = (A*x)/tau;
  
  //Construct sparce precision matrix for latent field---
  SparseMatrix<Type> Q;
  if(barrier==0){
    Q = Q_spde(spdeMatrices,kappa);
  }else if(barrier ==1){
    Q = Q_spde(spdeMatricesBarrier,kappa,c);
  }
  //---------------------------------------------

  //Calculates nll-------------------------------
  Type nll = 0.0;
  nll += GMRF(Q)(x);

  for(int i =0; i< y.size(); ++i){
    nll-= dnorm(xiid(i), Type(0), sigmaIID,true);
  }

  vector<Type> eta = exp(X*beta + delta + xiid);
  for(int i=0; i<y.size(); i++){
    nll -= dpois(y(i), eta(i),true);
  }
  //---------------------------------------------

  //Report what we want to report----------------
  Type range = sqrt(8)/kappa;   //Distance at which correlation has dropped to 0.1, see p. 4 in Lindgren et al. (2011)
  ADREPORT(range);
  //---------------------------------------------

  return nll;
}
